### ISSUE TITLE
gentoo image has empty repository path /var/db/repos/gentoo

### ISSUE THREAD

**Issue by pajot** [lxc/lxc-ci#184]
The base Gentoo image (966106a9d0cd) is missing the /var/db/repos/gentoo directory, which means that portage cannot be synced and the system is not modifiable by conventional means:

```
gentooC01 # emerge -av app-portage/eix
!!! Section 'gentoo' in repos.conf has location attribute set to nonexistent directory: '/var/db/repos/gentoo'
!!! Invalid Repository Location (not a dir): '/var/db/repos/gentoo'


!!! /etc/portage/make.profile is not a symlink and will probably prevent most merges.
!!! It should point into a profile within /var/db/repos/gentoo/profiles/
!!! (You can safely ignore this message when syncing. It's harmless.)


!!! Your current profile is invalid. If you have just changed your profile
!!! configuration, you should revert back to the previous configuration.
!!! Allowed actions are limited to --help, --info, --search, --sync, and
!!! --version.
gentooC01 # ls /var/db/repos/
gentooC01 # eselect profile list
!!! Error: Failed to get a list of valid profiles
exiting
gentooC01 # emerge --info
!!! Section 'gentoo' in repos.conf has location attribute set to nonexistent directory: '/var/db/repos/gentoo'
!!! Invalid Repository Location (not a dir): '/var/db/repos/gentoo'
WARNING: One or more repositories have missing repo_name entries:

        /var/db/repos/gentoo/profiles/repo_name

NOTE: Each repo_name entry should be a plain text file containing a
unique name for the repository on the first line.


Portage 2.3.103 (python 3.7.8-final-0, unavailable, gcc-9.3.0, glibc-2.30-r8, 4.19.97-gentoo x86_64)
=================================================================
System uname: Linux-4.19.97-gentoo-x86_64-Intel-R-_Core-TM-_i7-7600U_CPU_@_2.80GHz-with-gentoo-2.6
KiB Mem:    32804508 total,  23478692 free
KiB Swap:   32813020 total,  32813020 free
sh bash 5.0_p17
ld GNU ld (Gentoo 2.33.1 p2) 2.33.1
dev-lang/python:          2.7.18-r1::gentoo, 3.7.8-r2::gentoo, 3.8.4-r1::gentoo
sys-devel/autoconf:       2.69-r4::gentoo
sys-devel/automake:       1.16.1-r1::gentoo
sys-devel/binutils:       2.33.1-r1::gentoo
sys-devel/libtool:        2.4.6-r6::gentoo
sys-kernel/linux-headers: 5.4-r1::gentoo (virtual/os-headers)
Repositories:

ACCEPT_KEYWORDS="~amd64"
ACCEPT_LICENSE="* -@EULA"
CFLAGS="-O2 -pipe"
CONFIG_PROTECT="/etc /usr/share/gnupg/qualified.txt"
CONFIG_PROTECT_MASK="/etc/ca-certificates.conf /etc/env.d /etc/gentoo-release /etc/sandbox.d /etc/terminfo"
CXXFLAGS="-O2 -pipe"
DISTDIR="/var/cache/distfiles"
FEATURES="assume-digests binpkg-docompress binpkg-dostrip binpkg-logs config-protect-if-modified distlocks ebuild-locks fixlafiles ipc-sandbox merge-sync multilib-strict network-sandbox news parallel-fetch pid-sandbox preserve-libs protect-owned qa-unresolved-soname-deps sandbox sfperms strict unknown-features-warn unmerge-logs unmerge-orphans userfetch userpriv usersandbox usersync xattr"
GENTOO_MIRRORS="http://distfiles.gentoo.org"
PKGDIR="/var/cache/binpkgs"
PORTAGE_TMPDIR="/var/tmp"
USE=""
Unset:  CHOST, EMERGE_DEFAULT_OPTS, ENV_UNSET, PORTAGE_BINHOST, PORTAGE_BUNZIP2_COMMAND
gentooC01 # ls /usr/portage
ls: cannot access '/usr/portage': No such file or directory
gentooC01 # cat /usr/share/portage/config/repos.conf 
[DEFAULT]
main-repo = gentoo

[gentoo]
location = /var/db/repos/gentoo
sync-type = rsync
sync-uri = rsync://rsync.gentoo.org/gentoo-portage
auto-sync = yes
sync-rsync-verify-jobs = 1
sync-rsync-verify-metamanifest = yes
sync-rsync-verify-max-age = 24
sync-openpgp-key-path = /usr/share/openpgp-keys/gentoo-release.asc
sync-openpgp-keyserver = hkps://keys.gentoo.org
sync-openpgp-key-refresh-retry-count = 40
sync-openpgp-key-refresh-retry-overall-timeout = 1200
sync-openpgp-key-refresh-retry-delay-exp-base = 2
sync-openpgp-key-refresh-retry-delay-max = 60
sync-openpgp-key-refresh-retry-delay-mult = 4
sync-webrsync-verify-signature = yes
```

From the host:
```

$ lxc list -c n,f
+-----------+--------------+
|   NAME    |  BASE IMAGE  |
+-----------+--------------+
| gentooC01 | 966106a9d0cd |
+-----------+--------------+

```

```
# LANG=C lxc image list images:gentoo
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
|          ALIAS          | FINGERPRINT  | PUBLIC |               DESCRIPTION               | ARCHITECTURE |      TYPE       |   SIZE   |          UPLOAD DATE          |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
| gentoo (5 more)         | 24681eed9d52 | yes    | Gentoo current amd64 (20200731_16:07)   | x86_64       | VIRTUAL-MACHINE | 742.56MB | Jul 31, 2020 at 12:00am (UTC) |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
| gentoo (5 more)         | 966106a9d0cd | yes    | Gentoo current amd64 (20200731_16:07)   | x86_64       | CONTAINER       | 215.63MB | Jul 31, 2020 at 12:00am (UTC) |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
| gentoo/armhf (2 more)   | f8a29f4ca3fa | yes    | Gentoo current armhf (20200731_16:07)   | armv7l       | CONTAINER       | 182.80MB | Jul 31, 2020 at 12:00am (UTC) |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
| gentoo/cloud (3 more)   | 96a926a5abb2 | yes    | Gentoo current amd64 (20200731_17:11)   | x86_64       | VIRTUAL-MACHINE | 783.94MB | Jul 31, 2020 at 12:00am (UTC) |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
| gentoo/cloud (3 more)   | 97bdba229b88 | yes    | Gentoo current amd64 (20200731_17:11)   | x86_64       | CONTAINER       | 286.22MB | Jul 31, 2020 at 12:00am (UTC) |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
| gentoo/i386 (2 more)    | 72fd70dc7cf8 | yes    | Gentoo current i386 (20200731_16:07)    | i686         | CONTAINER       | 202.89MB | Jul 31, 2020 at 12:00am (UTC) |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
| gentoo/ppc64el (2 more) | f862fbef2770 | yes    | Gentoo current ppc64el (20200731_16:07) | ppc64le      | CONTAINER       | 204.23MB | Jul 31, 2020 at 12:00am (UTC) |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+
| gentoo/s390x (2 more)   | a0558a78b00c | yes    | Gentoo current s390x (20200731_16:07)   | s390x        | CONTAINER       | 186.47MB | Jul 31, 2020 at 12:00am (UTC) |
+-------------------------+--------------+--------+-----------------------------------------+--------------+-----------------+----------+-------------------------------+

```

**Comment by stgraber** [comment]
@monstermunchkin can you sort this out?

**Comment by pajot** [comment]
I've been trying to parse the gentoo.yaml script, and I can't see where /var/db/repos is being removed. In the [announcement for distrobuilder 1.0](https://discuss.linuxcontainers.org/t/distrobuilder-1-0-has-been-released/5987), Stéphane wrote

> #### Image sources
> 
> The image sources are distribution specific logic (in Go) to either retrieve a suitable minimal base image or produce one from scratch using commonly available tooling (like `debootstrap` ).

Is the base image incomplete, then? How can I define my own base image if I want to roll my own template?

**Comment by pajot** [comment]
Okay, I understand what's happening now. The image creates an environment corresponding [to this section of the Gentoo install process.](https://wiki.gentoo.org/wiki/Handbook:AMD64/Installation/Base#Configuring_Portage)

After running

`emerge --sync`

which succeeds, the repo directory is populated and emerge operations can proceed without errors.

The question is whether this should or should not be done when the image is generated. Is the intent to have a very bare-bones image and install any required packages using the yaml file?

The repository adds about 210 MB to the image, which is still reasonable, I think.

**Comment by stgraber** [comment]
I guess it was made to save space then. 210MB is quite a large increase on what's already one of our largest images. Combine that with the daily image updates and with how quickly that data becomes stale and I think it starts to make sense not to include it.

Though we'd probably want to compare the squashfs size with and without it to see what the actual impact is like

**Comment by pajot** [comment]
The existing gentoo VM image is already over 700 MB. Adding the repo the container image would make it about 430 MB.

Can we put the sync operation in an action that is run when the container instance is created? That way the image stays compact and the container is (more) ready.

**Comment by stgraber** [comment]
Yeah, I guess making it an init script that runs on first boot would work.

@monstermunchkin

**Comment by pajot** [comment]
> Yeah, I guess making it an init script that runs on first boot would work.

Do you mean like an OpenRC init script, or a configuration of lxd or lxc? If the learning curve isn't too steep on the latter, I can see about adding this, I've already forked the repo.
 
> Though we'd probably want to compare the squashfs size with and without it to see what the actual impact is like

I would expect that the repo would compress nicely, as it consists almost entirely of text files. (That 430 MB number assumes no compression of the repository at all.)

**Comment by stgraber** [comment]
I was thinking an openrc script which starts after networking is up and sync the data if it's not present already

**Comment by pajot** [comment]
Ah, ok, right, dead simple and practically obvious :)

I'll see what I can cook up.

**Comment by pajot** [comment]
I've made an attempt at a solution, see #185.

I cannot get the image to build (I cannot build an image with the original gentoo.yaml, either, I get two errors). This is working in Jenkins; what am I doing wrong?

**Comment by pajot** [comment]
Also, I noticed one other thing that's concerning. Why does the image have system-wide keywording? I'm referring to this:
https://github.com/lxc/lxc-ci/blob/5865c94b68256463d2bceaf32c9437d1cf059aa1/images/gentoo.yaml#L464

Unstable keywords for the entire system are not appropriate for production. If there are packages that need unstable versions, that is fine, but those packages need to be keyworded individually. We should assume that people will be using these images in production ("be generous in what you accept, be conservative in what you send").

**Comment by monstermunchkin** [comment]
I'm closing this issue as it has been resolved in #185 and #187.
